// sql-parser/lib/parser.js
// Generated by CoffeeScript 1.7.1
(function() {
  var buildParser;

  buildParser = function(opts) {
    var parser;
    parser = require('./compiled_parser').parser;
    opts = opts || {};
    parser.args = opts.args || [];
    parser.placeholders = opts.placeholders || [];
    parser.escFn = opts.escFn || function(v) {
      return v;
    };
    parser.lexer = {
      lex: function() {
        var tag, _ref;
        _ref = this.tokens[this.pos++] || [''], tag = _ref[0], this.yytext = _ref[1], this.yylineno = _ref[2];
        return tag;
      },
      setInput: function(tokens) {
        this.tokens = tokens;
        return this.pos = 0;
      },
      upcomingInput: function() {
        return "";
      }
    };
    parser.yy = require('./nodes');
    parser.throwUnMatchedParamsExc = function() {
      return parser.parseError("number of '?'s in statement string does not match argument count", {});
    };
    return parser;
  };

  exports.parser = buildParser();

  exports.throwUnMatchedParamsExc = exports.parser.throwUnMatchedParamsExc;

  exports.parse = function(str) {
    return buildParser().parse(str);
  };

  exports.parse2 = function(str, opts) {
    return buildParser(opts).parse(str);
  };

}).call(this);
